deflate.h

```diff
@@ -221,16 +222,27 @@ struct ALIGNED_(64) internal_state {
      * max_insert_length is used only for compression levels <= 6.
      */
 
-    int level;    /* compression level (1..9) */
-    int strategy; /* favor or force Huffman coding*/
+    int level;                  /* compression level (1..9) */
+    int strategy;               /* favor or force Huffman coding*/
+    unsigned int good_match;    /* Use a faster search when the previous match is longer than this */
+    int nice_match;             /* Stop searching when current match exceeds this */
+    unsigned int matches;       /* number of string matches in current block */
+    unsigned int insert;        /* bytes at end of window left to insert */
 
-    unsigned int good_match;
-    /* Use a faster search when the previous match is longer than this */
+    uint64_t bi_buf;            /* Output buffer.
+                                 * Bits are inserted starting at the bottom (least significant bits). */
+    int32_t bi_valid;           /* Number of valid bits in bi_buf.
+                                 * All bits above the last valid bit are always zero. */
 
-    int nice_match; /* Stop searching when current match exceeds this */
+    int heap_len;               /* number of elements in the heap */
+    int heap_max;               /* element of largest frequency */
+
+    int32_t padding1[1];
+
+                /* Cacheline 3 */
 
 #if defined(_M_IX86) || defined(_M_ARM)
-    int padding[2];
+    int32_t padding2[1];
 #endif
```

Comment:
Padding at start of cache line seems wrong... Padding should always be before alignment boundary.

Reply:
As far as I understand 32-bit architectures, those paddings are supposed to be after for example pointers to align before any struct members that are wider than 32 bits... It usually helps to put the struct members on a spreadsheet that contain bit widths of all the types for both 32-bit and 64-bit architectures.

Temporarily removing the padding should cause compiler to emit warning around the struct member that needs the padding. We have most warnings enabled and treated as errors on both Win32 and Win64 MSVC.

Reply:
Before the struct definition:

```
#ifdef _MSC_VER
#pragma warning(push)
#pragma warning(error : 4820)
#endif
```

and after the definition:

```
#ifdef _MSC_VER
#pragma warning(pop)
#endif
```

That should break the build if there is padding mismatch...